### ISSUE TITLE
[Question] Can fixtures be run in parallel?

### ISSUE THREAD

**Issue by anthony-legitfit** [microsoft/playwright#18577]
We're using fixtures to simulate the different types of users that can be logged in to our app. The time each fixture takes to set itself up isn't too long, however, it's common for us to use multiple fixtures in one test. So, with the fixtures getting set up in sequence, this time adds up. Is there a way to run our fixture code in parallel rather than in sequence?

For example, say we had the following fixtures:
```ts
export const test = base.extend<{
  businessPage: Page;
  customerPage: Page;
}>({
  businessPage: async ({ browser }, use) => {
    const context = await browser.newContext();
    const page = await context.newPage();
    // Assume this takes 1000ms.
    registerBusinessAccount(page);
    use(page);
    await context.close();
  },
  customerPage: async ({ browser }, use) => {
    const context = await browser.newContext();
    const page = await context.newPage();
    // Assume this takes 1000ms.
    registerCustomerAccount(page);
    use(page);
    await context.close();
  },
});
```
and we have the following test:
```ts
test('that customers can purchase from businesses', async ({
  businessPage,
  customerPage,
}) => {
  // ...
});
```
then the fixture set up would take ~2000ms to run. But if the fixture setup ran in parallel, then it would take ~1000ms to run.

Let me know if you need any more information. I've had a look around the docs but couldn't find anything relating to this, but I might have missed it!

**Comment by pavelfeldman** [comment]
#### Why was this issue closed?

Thank you for your involvement. This issue was closed due to limited engagement (upvotes/activity), lack of recent activity, and insufficient actionability. To maintain a manageable database, we prioritize issues based on these factors.

If you disagree with this closure, please open a new issue and reference this one. More support or clarity on its necessity may prompt a review. Your understanding and cooperation are appreciated.

**Comment by RenanGreca** [comment]
This has been closed for a while, but for future Googlers, here's one way to achieve it:

```js
export const test = base.extend<{
  pages: Page[];
}>({
  pages: async ({browser}, use) => {
    const pages = await Promise.all([
      test.step(`Launching Business Page`, async() => {
        const context = await browser.newContext();
        const page = await context.newPage();
        // Assume this takes 1000ms.
        registerBusinessAccount(page);
        return page;
      }),
      test.step(`Launching Business Page`, async() => {
        const context = await browser.newContext();
        const page = await context.newPage();
        // Assume this takes 1000ms.
        registerCustomerAccount(page);
        return page;
      })
    ])
    await use(pages)
    for (const page of pages) {
      await page.context().close();
    }
  }
});
```

Not the most elegant approach as it would bind the two fixtures together, but can be helpful to speed things up.

**Comment by mrozwat** [comment]
Для нас это тоже очень актуально. Мы иногда используем 3-5 контекстов и параллельный их запуск сильно бы ускорил наши тесты. Вот пример нашего кода

 ```
export const testWithAuth = base.extend<{
  anonymousPage: MainPage;
  baseLicensePage: MainPage;
  premiumLicensePage: MainPage;
  premiumLicensePageWithWebinars: MainPage;
  otherOrgLicensePage: MainPage;
  otherOrgLicensePageWithWebinars: MainPage;
  otherOrgLicensePageWithPremium: MainPage;
}>({
  anonymousPage: async ({ browser }, use, testInfo) => {
    await runUserContext({ browser, testInfo, use });
  },

  baseLicensePage: async ({ browser }, use, testInfo) => {
    const { username, password } = credentials.baseLicense[0];

    await runUserContext({ browser, testInfo, use, username, password });
  },

  premiumLicensePage: async ({ browser }, use, testInfo) => {
    const { username, password } = credentials.premLicense[0];

    await runUserContext({ browser, testInfo, use, username, password });
  },

  premiumLicensePageWithWebinars: async ({ browser }, use, testInfo) => {
    const { username, password } = credentials.premLicense.filter(
      (c) => c.features?.webinar,
    )[0];

    await runUserContext({ browser, testInfo, use, username, password });
  },

  otherOrgLicensePage: async ({ browser }, use, testInfo) => {
    const { username, password } = credentials.otherOrganization[0];

    await runUserContext({ browser, testInfo, use, username, password });
  },

  otherOrgLicensePageWithWebinars: async ({ browser }, use, testInfo) => {
    const { username, password } = credentials.otherOrganization.filter(
      (c) => c.features?.webinar,
    )[0];

    await runUserContext({ browser, testInfo, use, username, password });
  },

  otherOrgLicensePageWithPremium: async ({ browser }, use, testInfo) => {
    const { username, password } = credentials.otherOrganization.filter(
      (c) => !c.features?.webinar,
    )[1];

    await runUserContext({ browser, testInfo, use, username, password });
  },
});
```
